### ISSUE TITLE
Drop in ingestion rate when performing rolling update on vmstorage pods

### ISSUE THREAD

**Issue by ericholt** [VictoriaMetrics/VictoriaMetrics#7328]
### Describe the bug

When `vmstorage` pods are restarted one at a time during rolling updates, ingestion drops significantly (sometimes below 50%) and metrics accumulate in `vmagents` across all monitored k8s clusters and sometimes even dropped as the queue attains max capacity.

From analyzing the graphs and code, I've come to the conclusion that that while 1 `vmstorage` pod is shutting down its connections to `vminsert` pods, for a brief moment, the load is not evenly rerouted among the other `vmstorage` pods. Instead, the load is passed on to the next pod based on its sts pod id. For example, if `vmstorage-n` is down, `vmstorage-n+1` will get all its entire load. 

This is the sequence I believe is happening:

- `vmstorage` closes its connection with the `vminsert` pods
- `vminsert` doesn't yet realize this storage node's connection is closed
- `vminsert` keeps adding incoming metrics to its buffer dedicated to this storage node
- `vminsert` dedicated go routine for this buffer tries to send data to the no longer available storage node
  - falls back to the n+1 node when it realizes the connection is down (marks storage node as broken)
  - future incoming metrics get rerouted evenly to all storage nodes
- `vmstorage` node n+1 receives a large amount of "new" time series from every `vminsert` pod
  - it needs to update its indexes and cache causing a spike in CPU and memory usage
- `vminsert` reaches concurrency limit as it waits for `vmstorage`
  - incoming metrics are refused

There may be other ways of solving the ingestion issue, like increasing cpu and memory resources for each pod or some other configuration. While I am open to other solutions, there does seem to be a real problem caused by the sudden spike in incoming data to the `vmstorage-n+1` pod when the connections from `vmstorage-n` to all `vminsert` pods is closed.

### To Reproduce

We only need to perform a rolling update (1 pod at a time) of `vmstorage` pods.

We have the following `vmcluster` setup:

- 20 `vmstorage` pods
- 16  `vminsert` pods
- 16 `vmselect` pods
- Replication factor = 2

Monitoring stats:

- 19 monitored k8s clusters
- ~24 Tri datapoints
- ~3.6 Mil datapoints per second (7.2 million including replication factor)
- 114 Mil active time series (when stable)

### Version

Using cluster version `v1.101.0-cluster`
```
/ # ./vmstorage-prod --version
vmstorage-20240425-150523-tags-v1.101.0-cluster-0-g310d100ed
```

### Screenshots

Drop of ingestion rate during vmstorage rolling update:
![image-2024-10-18_17-2-0](https://github.com/user-attachments/assets/c83ef448-bff2-4c99-abc2-fb2d5f42ffe0)

Although vmstorage resource usage gets dangerously high, we don't get cpu throttled or OOM:
![image-2024-10-18_17-3-15](https://github.com/user-attachments/assets/e5f59258-2afe-4d18-84bd-5eb234f00df4)

We don't hit vmstorage max concurrency for writing to disk either:
![image-2024-10-18_17-3-49](https://github.com/user-attachments/assets/75be981a-41a4-4937-9404-308a1cad8b15)

But vminsert does hit max concurrency (set at 20, increasing to 40 did not resolve the problem):
![image-2024-10-18_17-5-17](https://github.com/user-attachments/assets/1d20c116-0286-4ba1-a701-59fece89011d)

Resource usage for vminsert, cpu usage suggests waiting on network connections, and memory usage suggests accumulation of metrics in buffers:
![image-2024-10-18_17-7-20](https://github.com/user-attachments/assets/3e82d7bc-6582-46d5-9bba-92c7ac7aac90)

Hitting max connection saturation time for vminsert as well:
![image-2024-10-18_17-7-54](https://github.com/user-attachments/assets/1f85b1e6-aabf-4c94-870e-88141a9d2c7f)

As `vmstorage-n` goes down, we get a cpu spike for `vmstorage-n+1` (only showing a few pods for clarity):
![image-2024-10-16_18-30-16](https://github.com/user-attachments/assets/2ac3ada6-c48b-447f-9d65-6dc98133d6d6)

### Used command-line flags

vmstorage
```
      args:
        - '-dedup.minScrapeInterval=5s'
        - '-httpListenAddr=:8482'
        - '-maxConcurrentInserts=24'
        - '-retentionPeriod=60d'
        - '-storageDataPath=/vm-data'
        - '-vminsertAddr=:8400'
        - '-vmselectAddr=:8401'
```

vminsert
```
      args:
        - '-httpListenAddr=:8480'
        - '-maxConcurrentInserts=20'
        - '-maxLabelsPerTimeseries=40'
        - '-replicationFactor=2'
        - >-
          -storageNode=<all 20 vmstorage pods>
```
I also tested with `-maxConcurrentInserts=40`

**Comment by ericholt** [comment]
Here are the graphs:
![image](https://github.com/user-attachments/assets/c7015de4-b742-4d4e-a266-401a64b64836)

The behaviour I describe above will not count toward this rerouting metric though as this is only counted once `vminsert` knows that a particular storage node's connection is closed.

**Comment by hagen1778** [comment]
> The behaviour I describe above will not count toward this rerouting metric though as this is only counted once vminsert knows that a particular storage node's connection is closed.

Indeed, it looks like you described. But it is unclear why it makes so big deal, because vminsert buffers are pretty small and rarely exceed couple of MBs. So this initial re-routing should cause significant impact on the cluster health.
@f41gh7 maybe you have different opinion?

**Comment by f41gh7** [comment]
The actual problem is configured replicationFactor > 1.  With current replication implementation it's not possible to evenly distribute load between storage nodes. Replication happens after possible rerouting.
  And if in normal case rerouting could help to utilize cluster compute resources by rerouting metrics from slow or saturated storage nodes. It's not working for enabled replication and it's expected to have ingestion slow down even at storage node restart. 

 Possible solutions:
1) perform minimal downtime strategy for updates: https://docs.victoriametrics.com/cluster-victoriametrics/#minimum-downtime-strategy
2) use "distributed" version of vmcluster deployment https://docs.victoriametrics.com/helm/victoriametrics-distributed/

 Related issue https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6801

**Comment by hagen1778** [comment]
@f41gh7 should we close it as a duplicate of https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6801 ?

**Comment by f41gh7** [comment]
Closing in favor of https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6801
